modules/images/image-loading-optimization/class-ilo-url-metric.php

```diff
@@ -0,0 +1,182 @@
+<?php
+/**
+ * Image Loading Optimization: ILO_URL_Metric class
+ *
+ * @package performance-lab
+ * @since n.e.x.t
+ */
+
+/**
+ * Representation of the measurements taken from a single client's visit to a specific URL.
+ *
+ * @since n.e.x.t
+ * @access private
+ */
+final class ILO_URL_Metric implements JsonSerializable {
+
+	/**
+	 * Data.
+	 *
+	 * @var array{
+	 *          timestamp: int,
+	 *          viewport: array{ width: int, height: int },
+	 *          elements: array<array{
+	 *              isLCP: bool,
+	 *              isLCPCandidate: bool,
+	 *              xpath: string,
+	 *              intersectionRatio: float,
+	 *              intersectionRect: array{ width: int, height: int },
+	 *              boundingClientRect: array{ width: int, height: int },
+	 *          }>
+	 *      }
+	 */
+	private $data;
+
+	/**
+	 * Constructor.
+	 *
+	 * @param array $data      URL metric data.
+	 * @param bool  $validated Whether the data was already validated.
+	 *
+	 * @throws Exception When the input is invalid.
+	 */
+	public function __construct( array $data, bool $validated = false ) {
```

Comment:
Yeah, this is the part I felt the most icky about. I did go down the route of having a subclass that doesn't require validation, but then thought it was maybe overkill to have a class just for this. I'm probably being too cautious here to avoid re-validating the data. I'll just make it require validation all the time.

Reply:
Done in c19c371e05d3c47eba2c39d0510cc1d1bb930a9a